// Copyright (c) 2025 Chris Lee and contibuters.
// Licensed under the MIT license. See LICENSE file in the project root for details.

syntax = "proto3";

import "nanopb.proto";

package og3;

message Sensor {
  uint32 id = 1;
  string name = 2 [ (nanopb).max_length = 15 ];
  string units = 3 [ (nanopb).max_length = 7 ];

  enum Type {
    TYPE_UNSPECIFIED = 0;
    TYPE_VOLTAGE = 1;
    TYPE_TEMPERATURE = 2;
    TYPE_HUMIDITY = 3;
    TYPE_MOISTURE = 4;
    TYPE_INT_NUMBER = 5;
  }
  Type type = 4;

  // State classes
  // https://developers.home-assistant.io/docs/core/entity/sensor/#available-state-classes
  enum StateClass {
    STATE_CLASS_UNSPECIFIED = 0;
    STATE_CLASS_MEASUREMENT = 1;
    // Not yet supported: MEASUREMENT_ANGLE, TOTAL, TOTAL_INCREASING
  }
  StateClass state_class = 5;
}

message FloatSensorReading {
  uint32 sensor_id = 1;
  float value = 2;
}

message IntSensorReading {
  uint32 sensor_id = 1;
  int32 value = 2;
}

message Version {
  uint32 major = 1;
  uint32 minor = 2;
  uint32 patch = 3;
}

message Device {
  uint32 id = 1;
  uint32 manufacturer = 2;
  string name = 3 [ (nanopb).max_length = 15 ];
  Version hardware_version = 4;
  Version software_version = 5;
}

message Packet {
  uint32 device_id = 1;
  Device device = 2;
  repeated FloatSensorReading reading = 3 [ (nanopb).max_length = 80, (nanopb).max_count = 8 ];
  repeated IntSensorReading i_reading = 4 [ (nanopb).max_length = 80, (nanopb).max_count = 8 ];
  repeated Sensor sensor = 5 [ (nanopb).max_length = 120, (nanopb).max_count = 8 ];
}
